src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/CertificateAuthorityService.java

```diff
@@ -0,0 +1,155 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.niis.xroad.restapi.service;
+
+import ee.ria.xroad.common.certificateprofile.CertificateProfileInfo;
+import ee.ria.xroad.common.certificateprofile.CertificateProfileInfoProvider;
+import ee.ria.xroad.common.certificateprofile.GetCertificateProfile;
+import ee.ria.xroad.common.certificateprofile.impl.AuthCertificateProfileInfoParameters;
+import ee.ria.xroad.common.certificateprofile.impl.SignCertificateProfileInfoParameters;
+import ee.ria.xroad.common.conf.globalconf.ApprovedCAInfo;
+import ee.ria.xroad.common.identifier.ClientId;
+import ee.ria.xroad.common.identifier.SecurityServerId;
+import ee.ria.xroad.signer.protocol.dto.KeyUsageInfo;
+
+import lombok.extern.slf4j.Slf4j;
+import org.niis.xroad.restapi.facade.GlobalConfFacade;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Service that handles approved certificate authorities
+ */
+@Slf4j
+@Service
+@Transactional
+@PreAuthorize("isAuthenticated()")
+public class CertificateAuthorityService {
+
+    private final GlobalConfService globalConfService;
+    private final ServerConfService serverConfService;
+    private final GlobalConfFacade globalConfFacade;
+    private final ClientService clientService;
+
+    /**
+     * constructor
+     */
+    @Autowired
+    public CertificateAuthorityService(GlobalConfService globalConfService,
+            GlobalConfFacade globalConfFacade,
+            ServerConfService serverConfService,
+            ClientService clientService) {
+        this.globalConfService = globalConfService;
+        this.globalConfFacade = globalConfFacade;
+        this.serverConfService = serverConfService;
+        this.clientService = clientService;
+    }
+
+    /**
+     * Return certificate authorities
+     * @param keyUsageInfo list CAs for this type of key usage. If null, list all.
+     * @return
+     */
+    public Collection<ApprovedCAInfo> getCertificateAuthorities(KeyUsageInfo keyUsageInfo) {
+        Collection<ApprovedCAInfo> matchingCas = globalConfService.getApprovedCAsForThisInstance();
+        if (keyUsageInfo == KeyUsageInfo.SIGNING) {
+            // need to remove "authentication only" CAs
+            matchingCas = matchingCas.stream()
+                    .filter(ca -> !(Boolean.TRUE.equals(ca.getAuthenticationOnly())))
+                    .collect(Collectors.toList());
```

Comment:
Do you mean like this:

```
        if (keyUsageInfo == KeyUsageInfo.SIGNING) {
            // need to remove "authentication only" CAs
            matchingCas.removeIf(ca -> !(Boolean.TRUE.equals(ca.getAuthenticationOnly())));
        }
``` ?

Good point, had not considered that! But, that would be different in that it has a side effect of potentially changing internal state of the object possibly holding matchingCas.

It turns out that would not happen, as `matchingCas` is a copy, but that's not too plain to see. You have to dig into globalConfService.getApprovedCAsForThisInstance() -> globalConfFacade.getApprovedCAs() -> GlobalConf.getApprovedCAs() -> getInstance().getApprovedCAs() -> go to concrete implementation at GlobalConfImpl.getApprovedCAs() to find the answer... And it's really not even a guaranteed thing, since API does not define such details, a valid (ok, a bad one) GlobalConf implementation might return the internal object.

Not sure that I consciously went through this thought process when writing those lines of code, but just wanted to expand.

Reply:
Did not change